cvsd/src/cvsd.c

```diff
@@ -36,112 +40,125 @@ short cvsdInit(cvsd_t *cvsd)
     return 0;
 }
 
-void cvsdEncode(cvsd_t *cvsd,  const short *restrict in, register unsigned int input_len,  unsigned int *restrict out)
+static inline void encode_bit(const int16_t *restrict *restrict in, register int32_t *restrict accum,
+        register int32_t *restrict step_size, register uint32_t *restrict output_byte)
 {
-    register unsigned int output_byte = cvsd->output_byte;
-    register int local_accum = cvsd->accumulator;
-    register int local_step_size = cvsd->step_size;
-
-    for (int i = 0; i < input_len; i += 32)
+    if ((*(*in)++ << PRECISION) >= *accum)
     {
-        for (int bit_counter = 0; bit_counter < 32; bit_counter++)
-        {
-            if ((*in++ << PRECISION) >= local_accum)
-            {
-                output_byte <<= 1;
-#ifdef __FXAPI__
-                local_accum = fx_add_q31(local_accum, local_step_size);
-#else
-                local_accum += local_step_size;
-                local_accum = _max(NEG_ACCUM_MAX, local_accum);
-#endif
-            }
-            else
-            {
-                output_byte = (output_byte << 1) | 1;
-#ifdef __FXAPI__
-                local_accum = fx_sub_q31(local_accum, local_step_size);
-#else
-                local_accum -= local_step_size;
-                local_accum = _min(POS_ACCUM_MAX, local_accum);
-#endif
-            }
-            local_accum -= (local_accum >> ETA_EXP);
-            unsigned int tmp = output_byte & BIT_MASK;
-            if ((tmp == BIT_MASK) || (tmp == 0))
-            {
-                local_step_size += MIN_DELTA;
-                local_step_size = _min(MAX_DELTA, local_step_size);
-            }
-            else
-            {
-                local_step_size -= (local_step_size >> BETA_EXP);
-                local_step_size = _max(MIN_DELTA, local_step_size);
-            }
-        }
+        *output_byte <<= 1;
+        *accum += *step_size;
+        *accum = _min(POS_ACCUM_MAX, *accum);
+    }
+    else
+    {
+        *output_byte = (*output_byte << 1) | 1;
+        *accum -= *step_size;
+        *accum = _max(NEG_ACCUM_MAX, *accum);
+    }
+    *accum -= (*accum >> ETA_EXP);
+    uint32_t tmp = *output_byte & BIT_MASK;
+    if ((tmp == BIT_MASK) || (tmp == 0))
+    {
+        *step_size += MIN_DELTA;
+        *step_size = _min(MAX_DELTA, *step_size);
+    }
+    else
+    {
+        *step_size -= (*step_size >> BETA_EXP);
+        *step_size = _max(MIN_DELTA, *step_size);
+    }
+}
+static inline void swap_endian(uint32_t * const restrict out_pos, const uint32_t *restrict output_byte)
```

Comment:
Why don't you use __builtin_bswap32() instead of this custom implementation? Also to my taste, too many pointers here. If somehow compiler does not inline this (e.g. -Os), performance will be bad.

Reply:
It is a bit slower (more 10 cycles per frame) but if it better approach to use builtins I will use that function.
About -Os is slower if I don't use function encode_bit, but I add **__attribute__((always_inline))** it improved performance for Os.

Reply:
Sorry, my previous experiments with __builtin_bswap32() were incorrect.
We have not any difference between using the intrinsic and builtin function.